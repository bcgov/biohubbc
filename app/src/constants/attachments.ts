export enum AttachmentType {
  REPORT = 'Report',
  OTHER = 'Other'
}

export enum ProjectAttachmentValidExtensions {
  AUDIO = '.wav, .mp3, .mp4, .wma',
  DATA = '.txt, .xls, .xlsx, .xlsm, .xlsb, .accdb, .mdb, .ods, .csv',
  IMAGE = '.gif, .png, .jpg, .jpeg, .svg, .tiff, .bmp, .tif',
  REPORT = '.doc, .docx, .pdf',
  SPATIAL = '.kml, .gpx, .zip',
  VIDEO = '.mp4, .mov, .wmv, .ave',
  OTHER = '.wav, .mp3, .mp4, .wma, .txt, .xls, .xlsx, .xlsm, .xlsb, .accdb, .mdb, .ods, .csv, .gif, .png, .jpg, .jpeg, .svg, .tiff, .bmp, .tif, .doc, .docx, .kml, .gdb, .lyr, .shp, .las, .laz, .zlas, .gml, .gpx, .json, .mp4, .mov, .wmv, .ave'
}
