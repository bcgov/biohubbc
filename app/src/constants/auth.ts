export enum SYSTEM_IDENTITY_SOURCE {
  BCEID_BUSINESS = 'BCEIDBUSINESS',
  BCEID_BASIC = 'BCEIDBASIC',
  IDIR = 'IDIR',
  DATABASE = 'DATABASE',
  UNVERIFIED = 'UNVERIFIED'
}
