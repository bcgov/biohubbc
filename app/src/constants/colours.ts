import { Color } from '@mui/material';
import {
  blue,
  blueGrey,
  brown,
  cyan,
  deepOrange,
  deepPurple,
  green,
  grey,
  indigo,
  orange,
  pink,
  purple,
  red,
  teal
} from '@mui/material/colors';

type ColourMap = Record<string, { colour: Color }>;

/**
 * Default colour fallback.
 *
 */
const DEFAULT_COLOUR = blueGrey;

/**
 * Colour map for `Survey Progress` chips.
 *
 */
const SURVEY_PROGRESS_COLOUR_MAP = {
  Planning: { colour: blueGrey },
  'In progress': { colour: deepPurple },
  Completed: { colour: green }
};

/**
 * Colour map for `Taxon Rank` chips.
 *
 */
const TAXON_RANK_COLOUR_MAP = {
  Subspecies: { colour: blue },
  Variety: { colour: blue },
  Species: { colour: purple },
  Genus: { colour: teal },
  Family: { colour: red },
  Order: { colour: indigo },
  Class: { colour: deepOrange },
  Phylum: { colour: pink },
  Kingdom: { colour: grey }
};

/**
 * Colour map for `NRM Region` chips.
 *
 */
const NRM_REGION_COLOUR_MAP = {
  'Kootenay-Boundary Natural Resource Region': { colour: cyan },
  'Thompson-Okanagan Natural Resource Region': { colour: orange },
  'West Coast Natural Resource Region': { colour: green },
  'Cariboo Natural Resource Region': { colour: deepPurple },
  'South Coast Natural Resource Region': { colour: blue },
  'Northeast Natural Resource Region': { colour: brown },
  'Omineca Natural Resource Region': { colour: pink },
  'Skeena Natural Resource Region': { colour: red }
};

/**
 * Generate colour getter from ColourMap.
 *
 * @param {ColourMap} colourMap
 * @param {Color} [fallbackColour]
 * @returns {(lookup: string) => Color} Colour getter
 */
const generateColourMapGetter = <T extends ColourMap>(colourMap: T, fallbackColour = DEFAULT_COLOUR) => {
  return (lookup: string) => colourMap[lookup]?.colour ?? fallbackColour;
};

/**
 * Get survey progress colour mapping.
 *
 */
export const getSurveyProgressColour = generateColourMapGetter(SURVEY_PROGRESS_COLOUR_MAP);

/**
 * Get taxon rank colour mapping.
 *
 */
export const getTaxonRankColour = generateColourMapGetter(TAXON_RANK_COLOUR_MAP);

/**
 * Get NRM region colour mapping.
 *
 */
export const getNrmRegionColour = generateColourMapGetter(NRM_REGION_COLOUR_MAP);
