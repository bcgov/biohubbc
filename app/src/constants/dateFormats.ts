// See BC Gov standards: https://www2.gov.bc.ca/gov/content/governments/services-for-government/policies-procedures/web-content-development-guides/writing-for-the-web/web-style-guide/numbers

export const ShortDateFormat = 'YYYY-MM-DD'; //2020-01-05

export const ShortDateTimeFormat = 'YYYY-MM-DD, H:mm a'; //2020-01-05, 3:30 pm

export const MediumDateFormat = 'MMMM D, YYYY'; //January 5, 2020

export const MediumDateTimeFormat = 'MMMM D, YYYY, H:mm a'; //January 5, 2020, 3:30 pm

export const LongDateFormat = 'dddd, MMMM D, YYYY, H:mm a'; //Monday, January 5, 2020, 3:30 pm

export const LongDateTimeFormat = 'dddd, MMMM D, YYYY, H:mm a'; //Monday, January 5, 2020, 3:30 pm
