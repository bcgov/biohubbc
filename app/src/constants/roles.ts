export const Roles = {
  SYSTEM_ADMINISTRATOR: 'ADM'
};
