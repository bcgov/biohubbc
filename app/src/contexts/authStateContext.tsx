import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import React from 'react';

export interface IAuthState {
  ready?: boolean;
}

export const AuthStateContext = React.createContext<IAuthState>({
  ready: false
});

export const AuthStateContextProvider: React.FC = (props) => {
  const keycloakWrapper = useKeycloakWrapper();

  return (
    <AuthStateContext.Provider value={{ ready: keycloakWrapper.keycloak?.authenticated }}>
      {props.children}
    </AuthStateContext.Provider>
  );
};
