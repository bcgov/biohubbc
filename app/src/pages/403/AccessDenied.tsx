import { Container, Grid } from '@material-ui/core';
import React from 'react';
import { Link } from 'react-router-dom';

const AccessDenied = () => {
  return (
    <Container>
      <Grid container direction="row">
        <Grid item>
          <h1>Access Denied</h1>
          <h2>You do not have permission to view this page</h2>
          <Link to="/home">Go back to the home page</Link>
        </Grid>
      </Grid>
    </Container>
  );
};

export default AccessDenied;
