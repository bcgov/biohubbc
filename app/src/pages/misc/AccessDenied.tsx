import * as React from 'react';
import { Link } from 'react-router-dom';
import { Container, Grid } from '@material-ui/core';

const AccessDenied = () => {
  return (
    <Container>
      <Grid container direction="row">
        <Grid item>
          <h1>Access Denied</h1>
          <h2>You do not have permission to view this page</h2>
          <Link to="/home/activities">Go back to the activities page</Link>
        </Grid>
      </Grid>
    </Container>
  );
};

export default AccessDenied;
