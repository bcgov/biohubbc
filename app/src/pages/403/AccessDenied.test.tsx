import { fireEvent, render } from '@testing-library/react';
import { SYSTEM_ROLE } from 'constants/roles';
import { AuthStateContext } from 'contexts/authStateContext';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Router } from 'react-router-dom';
import AccessDenied from './AccessDenied';

const history = createMemoryHistory();

describe('AccessDenied', () => {
  it('redirects to `/` when user is not authenticated', () => {
    const authState = {
      keycloakWrapper: {
        keycloak: {
          authenticated: false
        },
        hasLoadedAllUserInfo: false,
        hasAccessRequest: false,

        systemRoles: [],
        getUserIdentifier: jest.fn(),
        hasSystemRole: jest.fn(),
        getIdentitySource: jest.fn(),
        username: 'testusername',
        displayName: 'testdisplayname',
        email: 'test@email.com',
        firstName: 'testfirst',
        lastName: 'testlast',
        refresh: () => {}
      }
    };

    const history = createMemoryHistory();

    history.push('/forbidden');

    render(
      <AuthStateContext.Provider value={authState}>
        <Router history={history}>
          <AccessDenied />
        </Router>
      </AuthStateContext.Provider>
    );

    expect(history.location.pathname).toEqual('/');
  });

  it('renders a spinner when user is authenticated and `hasLoadedAllUserInfo` is false', () => {
    const authState = {
      keycloakWrapper: {
        keycloak: {
          authenticated: true
        },
        hasLoadedAllUserInfo: false,
        hasAccessRequest: false,

        systemRoles: [],
        getUserIdentifier: jest.fn(),
        hasSystemRole: jest.fn(),
        getIdentitySource: jest.fn(),
        username: 'testusername',
        displayName: 'testdisplayname',
        email: 'test@email.com',
        firstName: 'testfirst',
        lastName: 'testlast',
        refresh: () => {}
      }
    };

    const history = createMemoryHistory();

    history.push('/forbidden');

    const { asFragment } = render(
      <AuthStateContext.Provider value={authState}>
        <Router history={history}>
          <AccessDenied />
        </Router>
      </AuthStateContext.Provider>
    );

    // does not change location
    expect(history.location.pathname).toEqual('/forbidden');

    // renders a spinner
    expect(asFragment()).toMatchSnapshot();
  });

  it('redirects to `/request-submitted` when user is authenticated and has a pending access request', () => {
    const authState = {
      keycloakWrapper: {
        keycloak: {
          authenticated: true
        },
        hasLoadedAllUserInfo: true,
        hasAccessRequest: true,

        systemRoles: [],
        getUserIdentifier: jest.fn(),
        hasSystemRole: jest.fn(),
        getIdentitySource: jest.fn(),
        username: 'testusername',
        displayName: 'testdisplayname',
        email: 'test@email.com',
        firstName: 'testfirst',
        lastName: 'testlast',
        refresh: () => {}
      }
    };

    const history = createMemoryHistory();

    history.push('/forbidden');

    render(
      <AuthStateContext.Provider value={authState}>
        <Router history={history}>
          <AccessDenied />
        </Router>
      </AuthStateContext.Provider>
    );

    expect(history.location.pathname).toEqual('/request-submitted');
  });

  it('renders correctly when the user is authenticated and has no pending access requests', () => {
    const authState = {
      keycloakWrapper: {
        keycloak: {
          authenticated: true
        },
        hasLoadedAllUserInfo: true,
        hasAccessRequest: false,

        systemRoles: [SYSTEM_ROLE.PROJECT_CREATOR],
        getUserIdentifier: jest.fn(),
        hasSystemRole: jest.fn(),
        getIdentitySource: jest.fn(),
        username: 'testusername',
        displayName: 'testdisplayname',
        email: 'test@email.com',
        firstName: 'testfirst',
        lastName: 'testlast',
        refresh: () => {}
      }
    };

    const { getByText, queryByTestId } = render(
      <AuthStateContext.Provider value={authState}>
        <Router history={history}>
          <AccessDenied />
        </Router>
      </AuthStateContext.Provider>
    );

    expect(getByText('You do not have permission to access this page.')).toBeVisible();
    expect(queryByTestId('request_access')).not.toBeInTheDocument();
  });

  it('redirects to `/access-request` when the `Request Access` button clicked', () => {
    const authState = {
      keycloakWrapper: {
        keycloak: {
          authenticated: true
        },
        hasLoadedAllUserInfo: true,
        hasAccessRequest: false,

        systemRoles: [],
        getUserIdentifier: jest.fn(),
        hasSystemRole: jest.fn(),
        getIdentitySource: jest.fn(),
        username: 'testusername',
        displayName: 'testdisplayname',
        email: 'test@email.com',
        firstName: 'testfirst',
        lastName: 'testlast',
        refresh: () => {}
      }
    };

    const { getByText, getByTestId } = render(
      <AuthStateContext.Provider value={authState}>
        <Router history={history}>
          <AccessDenied />
        </Router>
      </AuthStateContext.Provider>
    );

    expect(getByText('You do not have permission to access this application.')).toBeVisible();
    expect(getByTestId('request_access')).toBeVisible();

    fireEvent.click(getByText('Request Access'));

    expect(history.location.pathname).toEqual('/access-request');
  });
});
