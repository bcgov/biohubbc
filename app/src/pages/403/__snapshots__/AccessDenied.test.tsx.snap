// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotFoundPage renders correctly 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
>
  <div
    className="MuiGrid-root MuiGrid-container"
  >
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <h1>
        Access Denied
      </h1>
      <h2>
        You do not have permission to view this page
      </h2>
      <a
        href="/home"
        onClick={[Function]}
      >
        Go back to the home page
      </a>
    </div>
  </div>
</div>
`;
